lib/system/iterators.nim

```diff
@@ -129,14 +129,141 @@ iterator items*[T: Ordinal](s: Slice[T]): T =
   for x in s.a .. s.b:
     yield x
 
+iterator ritems*[T: not char](a: openArray[T]): lent2 T {.inline.} =
+  ## Iterates over each item of `a` in reverse.
+  var i = high(a)
+  while i >= low(a):
+    yield a[i]
+    dec(i)
+
+iterator ritems*[T: char](a: openArray[T]): T {.inline.} =
+  ## Iterates over each item of `a` in reverse.
+  # a VM bug currently prevents taking address of openArray[char]
+  # elements converted from a string (would fail in `tests/misc/thallo.nim`)
+  # in any case there's no performance advantage of returning char by address.
+  var i = high(a)
+  while i >= low(a):
+    yield a[i]
+    dec(i)
+
+iterator mritems*[T](a: var openArray[T]): var T {.inline.} =
+  ## Iterates over each item of `a` in reverse so that the yielded value can be modified.
+  var i = high(a)
+  while i >= low(a):
+    yield a[i]
+    dec(i)
+
+iterator ritems*[IX, T](a: array[IX, T]): T {.inline.} =
+  ## Iterates over each item of `a` in reverse.
+  when a.len > 0:
+    var i = high(IX)
+    while true:
+      yield a[i]
+      if i < low(IX): break
+      dec(i)
+
+iterator mritems*[IX, T](a: var array[IX, T]): var T {.inline.} =
+  ## Iterates over each item of `a`in reverse so that the yielded value can be modified.
+  when a.len > 0:
+    var i = high(IX)
+    while true:
+      yield a[i]
+      if i < low(IX): break
+      dec(i)
+
+iterator ritems*[T](a: set[T]): T {.inline.} =
+  ## Iterates over each element of `a` in reverse. `items` iterates only over the
+  ## elements that are really in the set (and not over the ones the set is
+  ## able to hold).
+  var i = high(T).int
+  while i > low(T).int:
+    if T(i) in a: yield T(i)
+    dec(i)
+
+iterator ritems*(a: cstring): char {.inline.} =
+  ## Iterates over each item of `a` in reverse.
+  runnableExamples:
+    from std/sequtils import toSeq
+    assert toSeq("abc\0def".cstring) == @['a', 'b', 'c']
+    assert toSeq("abc".cstring) == @['a', 'b', 'c']
```

Comment:
If you want to keep the examples, they should use `ritems`. Same for the other `runnableExamples`.

Reply:
Done